module/surface_sam_vhf.c

```diff
@@ -117,23 +683,83 @@ static void vhf_hid_close(struct hid_device *hid)
 
 static int vhf_hid_parse(struct hid_device *hid)
 {
-	return hid_parse_report(hid, (u8 *)vhf_hid_desc, ARRAY_SIZE(vhf_hid_desc));
+	struct vhf_drvdata *data = platform_get_drvdata(to_platform_device(hid->dev.parent));
+
+	switch (data->event_ctx.info->type) {
+	case VHF_MSHW0096:
+		return hid_parse_report(hid, (u8 *)vhf_hid_keyboard_only_desc, ARRAY_SIZE(vhf_hid_keyboard_only_desc));
+	case VHF_MSHW0114:
+		return hid_parse_report(hid, (u8 *)vhf_hid_desc, ARRAY_SIZE(vhf_hid_desc));
+	};
+
+	return -EINVAL;
 }
 
 static int vhf_hid_raw_request(struct hid_device *hid, unsigned char reportnum,
 			       u8 *buf, size_t len, unsigned char rtype,
 			       int reqtype)
 {
-	hid_dbg(hid, "%s\n", __func__);
-	return 0;
-}
+	// Byte 0 is the report number. Report data starts at byte 1.
+	buf[0] = reportnum;
 
-static int vhf_hid_output_report(struct hid_device *hid, u8 *buf, size_t len)
-{
-	hid_dbg(hid, "%s\n", __func__);
+	hid_dbg(hid, "%s: reportnum %#04x %i %i\n", __func__, reportnum, rtype, reqtype);
 	print_hex_dump_debug("report:", DUMP_PREFIX_OFFSET, 16, 1, buf, len, false);
 
-	return len;
+	int status;
+	u8 iid = 0x02;
+	u8 cid;
+
+	struct vhf_drvdata *data = platform_get_drvdata(to_platform_device(hid->dev.parent));
+
+	switch (reqtype) {
+	case HID_REQ_GET_REPORT:
+		cid = 0x02;
+		break;
+	case HID_REQ_SET_REPORT:
+		cid = 0x03;
+		break;
+	default:
+		hid_err(hid, "%s: unknown req type 0x%02x\n", __func__, rtype);
+		return 0;		// discard everything
+	}
+
+	// HAXX: The keyboard and the touchpad should probably be separate drivers.
```

Comment:
Not too happy about this, we'd need to probably have two drivers (and then we can just load the keyboard one for both devices, the touchpad one for the new one), but there's no way to split an event stream based on the IID.

Reply:
This should be resolved for now, using IID 0. We might consider splitting up the devices by IID to replicate the behavior on Windows more closely as a follow-up.